examples/java/src/test/java/org/apache/beam/examples/complete/kafkatopubsub/KafkaToPubsubTest.java

```diff
@@ -20,20 +20,108 @@
+import org.testcontainers.utility.DockerImageName;
 
 /** Test of KafkaToPubsub. */
 @RunWith(JUnit4.class)
 public class KafkaToPubsubTest {
+  @Rule
+  public final transient TestPipeline pipeline = TestPipeline.fromOptions(OPTIONS);
+  @Rule
+  public transient TestPubsubSignal signal = TestPubsubSignal.fromOptions(OPTIONS);
+
+  private static final String PUBSUB_EMULATOR_IMAGE = "gcr.io/google.com/cloudsdktool/cloud-sdk:316.0.0-emulators";
+  private static final String PUBSUB_MESSAGE = "test pubsub message";
+  private static final String PROJECT_ID = "try-kafka-pubsub";
+  private static final String TOPIC_NAME = "listen-to-kafka";
+  private static final PubsubClient.TopicPath TOPIC_PATH = PubsubClient
+          .topicPathFromName(PROJECT_ID, TOPIC_NAME);
+  private static final PipelineOptions OPTIONS = TestPipeline.testingPipelineOptions();
+
+  @BeforeClass
+  public static void beforeClass() throws Exception {
+    Credentials credentials = NoopCredentialFactory.getInstance().getCredential();
+    OPTIONS.as(GcpOptions.class).setGcpCredential(credentials);
+    OPTIONS.as(GcpOptions.class).setProject(PROJECT_ID);
+    setupPubsubContainer(OPTIONS.as(PubsubOptions.class));
+    createPubsubTopicForTest(OPTIONS.as(PubsubOptions.class));
+  }
+
+  @Test
+  public void testKafkaToPubsubE2E() throws IOException {
+    pipeline.getOptions().as(DirectOptions.class).setBlockOnRun(false);
+
+    RunKafkaContainer rkc = new RunKafkaContainer(PUBSUB_MESSAGE);
+    String bootstrapServer = rkc.getBootstrapServer();
+    String[] kafkaTopicsList = new String[]{rkc.getTopicName()};
+
+    String pubsubTopicPath = TOPIC_PATH.getPath();
+
+    Map<String, Object> kafkaConfig = new HashMap<>();
+    Map<String, String> sslConfig = new HashMap<>();
+
+    PCollection<KV<String, String>> readStrings = pipeline
+        .apply("readFromKafka",
+            readFromKafka(bootstrapServer, Arrays.asList(kafkaTopicsList), kafkaConfig, sslConfig));
+
+    PCollection<String> readFromPubsub = readStrings.apply(Values.create())
+        .apply("writeToPubSub", PubsubIO.writeStrings().to(pubsubTopicPath)).getPipeline()
+        .apply("readFromPubsub",
+            PubsubIO.readStrings().fromTopic(pubsubTopicPath));
+
+    readFromPubsub.apply("waitForTestMessage", signal.signalSuccessWhen(
+                readFromPubsub.getCoder(),
+                input -> {
+                  if (input == null) return false;
+                  return input.stream().anyMatch(message -> Objects.equals(message, PUBSUB_MESSAGE));
+                })
+    );
+
+    Supplier<Void> start = signal.waitForStart(Duration.standardSeconds(10));
+    pipeline.apply(signal.signalStart());
+    PipelineResult job = pipeline.run();
+    start.get();
+    signal.waitForSuccess(Duration.standardMinutes(2));
+    try {
+      job.cancel();
+    } catch (IOException | UnsupportedOperationException e) {
+      System.out.println("Something went wrong");
```

Comment:
Shouldn't we use logging for this (I, actually, don't know much about logging in beam tests)? And let's provide something more meaningful to describe the reason for cancelation failure, it would be also good to add an exception that was caught.

Reply:
Added throwing of `AssertionError` here